src/Control/Eff/State/OnDemand.hs

```diff
@@ -71,47 +71,46 @@ put s = send (Put s)
 onDemand :: Member (OnDemandState s) r => Eff '[OnDemandState s] v -> Eff r v
 onDemand = send . Delay
 
-runState' :: Eff (OnDemandState s ': r) w -> s -> Eff r (w,s)
-runState' m s =
+runState' :: s -> Eff (OnDemandState s ': r) w -> Eff r (w,s)
+runState' s =
   handle_relay_s s
   (\s0 x -> return (x,s0))
   (\s0 sreq k -> case sreq of
       Get    -> k s0 s0
       Put s1 -> k s1 ()
-      Delay m1 -> let ~(x,s1) = run $ runState' m1 s0
+      Delay m1 -> let ~(x,s1) = run $ runState' s0 m1
                   in k s1 x)
-  m
 
 -- Since State is so frequently used, we optimize it a bit
 -- | Run a State effect
-runState :: Eff (OnDemandState s ': r) w -- ^ Effect incorporating State
-         -> s                    -- ^ Initial state
-         -> Eff r (w,s)          -- ^ Effect containing final state and a return value
-runState (Val x) s = return (x,s)
-runState (E u0 q) s0 = case decomp u0 of
-  Right Get     -> runState (q ^$ s0) s0
-  Right (Put s1) -> runState (q ^$ ()) s1
-  Right (Delay m1) -> let ~(x,s1) = run $ runState m1 s0
-                      in runState (q ^$ x) s1
-  Left  u -> E u (singleK (\x -> runState (q ^$ x) s0))
+runState :: s                            -- ^ Initial state
+         -> Eff (OnDemandState s ': r) w -- ^ Effect incorporating State
+         -> Eff r (w,s)                  -- ^ Effect containing final state and a return value
+runState s (Val x) = return (x,s)
+runState s0 (E u0 q) = case decomp u0 of
+  Right Get     -> runState s0 (q ^$ s0)
+  Right (Put s1) -> runState s1 (q ^$ ())
+  Right (Delay m1) -> let ~(x,s1) = run $ runState s0 m1
+                      in runState s1 (q ^$ x)
+  Left  u -> E u (singleK (\x -> runState s0 (q ^$ x)))
 
 -- | Transform the state with a function.
 modify :: (Member (OnDemandState s) r) => (s -> s) -> Eff r ()
 modify f = get >>= put . f
 
 -- | Run a State effect, discarding the final state.
-evalState :: Eff (OnDemandState s ': r) w -> s -> Eff r w
-evalState m s = fmap fst . flip runState s $ m
+evalState :: s -> Eff (OnDemandState s ': r) w -> Eff r w
+evalState s = fmap fst . runState s
 
 -- | Run a State effect and return the final state.
-execState :: Eff (OnDemandState s ': r) w -> s -> Eff r s
-execState m s = fmap snd . flip runState s $ m
+execState :: s -> Eff (OnDemandState s ': r) w -> Eff r s
+execState s = fmap snd . runState s
 
 -- | A different representation of State: decomposing State into mutation
 -- (Writer) and Reading. We don't define any new effects: we just handle the
 -- existing ones.  Thus we define a handler for two effects together.
-runStateR :: Eff (Writer s ': Reader s ': r) w -> s -> Eff r (w,s)
-runStateR m0 s0 = loop s0 m0
+runStateR :: s -> Eff (Writer s ': Reader s ': r) w -> Eff r (w,s)
+runStateR s m = loop s m
```

Comment:
@power-fungus please fix the name shadowing error. i believe this is triggeringit

Reply:
This should be indeed the issue. However, using `stack build` this does not get detected. Is there a way to reduce the complexity of the Makefile by using stack? stack builds is fine, but does not report the shadowing errors. I don't understand the Makefile well enough to transform that. `make build` fails on my machine with

```
cabal sandbox init
Writing a default package environment file to
/home/power-fungus/Documents/haskell/extensible-effects/cabal.sandbox.config
Using an existing sandbox located at
/home/power-fungus/Documents/haskell/extensible-effects/.cabal-sandbox
cabal clean
cleaning...
cabal install --only-dependencies --enable-tests --enable-benchmarks
Warning: The package list for 'hackage.haskell.org' does not exist. Run 'cabal
update' to download it.
Resolving dependencies...
cabal: Could not resolve dependencies:
rejecting: extensible-effects-2.5.2.0:!bench (constraint from config file,
command line flag, or user target requires opposite flag selection)
trying: extensible-effects-2.5.2.0:*bench
unknown package: test-framework-th (dependency of
extensible-effects-2.5.2.0:*bench)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: extensible-effects,
extensible-effects-2.5.2.0:bench, test-framework-th
Note: when using a sandbox, all packages are required to have consistent
dependencies. Try reinstalling/unregistering the offending packages or
recreating the sandbox.
```

Reply:
Also, it is possible to travis-cache the dependencies (which I know how to do for stack) but not for cabal with sandboxes

Reply:
@power-fungus a couple of points:
- `make build` will fail unless a particular version of `cabal` and `ghc` are in scope. when developing, i bring those in via `nix` by running `nix-shell` (or a variant)
- happy to move `make` build steps to use `stack` instead. patches welcome! (same for `travis-cache` issue)